packages/playwright-crawler/src/internals/adaptive-playwright-crawler.ts

```diff
@@ -0,0 +1,253 @@
+import { cheerioPortadom, playwrightLocatorPortadom, type Portadom } from 'portadom';
+
+import type { PlaywrightCrawlerOptions, PlaywrightCrawlingContext } from './playwright-crawler';
+import { PlaywrightCrawler } from './playwright-crawler';
+import { RenderingTypePredictor, type RenderingType } from './utils/rendering-type-prediction';
+
+type Result<TResult> = {result: TResult; ok: true} | {error: unknown; ok: false}
+
+interface AdaptivePlaywrightCrawlerContext extends RestrictedCrawlingContext {
+    dom: Portadom<unknown, unknown>;
+    querySelector: (selector: string, timeoutMs?: number) => Awaitable<Cheerio<Element>>;
+}
+
+export interface AdaptivePlaywrightCrawlerOptions extends Omit<PlaywrightCrawlerOptions, 'requestHandler'> {
+    /**
+     * Function that is called to process each request.
+     *
+     * The function receives the {@apilink AdaptivePlaywrightCrawlingContext} as an argument, and it must refrain from calling code with side effects,
+     * other than the methods of the crawling context. Any other side effects may be invoked repeatedly by the crawler, which can lead to inconsistent results.
+     *
+     * The function must return a promise, which is then awaited by the crawler.
+     *
+     * If the function throws an exception, the crawler will try to re-crawl the
+     * request later, up to `option.maxRequestRetries` times.
+     */
+    requestHandler: (crawlingContext: AdaptivePlaywrightCrawlerContext) => Awaitable<void>;
+
+    /**
+     * Specifies the frequency of rendering type detection checks - 0.1 means roughly 10% of requests.
+     */
+    renderingTypeDetectionRatio: number;
+
+    /**
+     * An optional callback that is called on dataset items found by the request handler in plain HTTP mode.
+     * If it returns false, the request is retried in a browser.
+     * If no callback is specified, every dataset item is considered valid.
+     */
+    resultChecker?: (result: RequestHandlerResult) => boolean;
+
+    /**
+     * An optional callback used in rendering type detection. On each detection, the result of the plain HTTP run is compared to that of the browser one.
+     * If the callback returns true, the results are considered equal and the target site is considered static.
+     * If no result comparator is specified, but there is a `resultChecker`, any site where the `resultChecker` returns true is considered static.
+     * If neither `resultComparator` nor `resultChecker` are specified, a deep comparison of returned dataset items is used as a default.
+     */
+    resultComparator?: (resultA: RequestHandlerResult, resultB: RequestHandlerResult) => boolean;
+
+    /**
+     * A custom rendering type predictor
+     */
+    renderingTypePredictor?: Pick<RenderingTypePredictor, 'predict' | 'storeResult'>;
+}
+
+/**
+ * An extension of {@apilink PlaywrightCrawler} that uses a more limited interface so that it is able to switch to HTTP-only crawling when it detects it may be possible.
```

Comment:
I know this is very succint. Let me extend that once we're in sync on how the DOM querying API will work.